spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -0,0 +1,152 @@
+                                               (implicit oldTag: ClassTag[O], val rct: ClassTag[N],
+                                                @transient val rwf: RowWriterFactory[O], @transient val rtf: RowReaderFactory[N])
+  extends CassandraRDD[(O, N)](prev.sparkContext, prev.dependencies) with CassandraReader[N] {
+
+  //Make sure copy operations make new CJRDDs and not CRDDs
+  override protected def copy(columnNames: ColumnSelector = columnNames,
+                              where: CqlWhereClause = where,
+                              readConf: ReadConf = readConf, connector: CassandraConnector = connector) =
+    new CassandraJoinRDD[O, N](prev, keyspaceName, tableName, connector, columnNames, joinColumns, where, readConf).asInstanceOf[this.type]
+
+  lazy val joinColumnNames: Seq[NamedColumnRef] = joinColumns match {
+    case AllColumns => throw new IllegalArgumentException("Unable to join against all columns in a Cassandra Table. Only primary key columns allowed")
+    case PartitionKeyColumns => tableDef.partitionKey.map(col => col.columnName: NamedColumnRef).toSeq
+    case SomeColumns(cs@_*) => {
+      checkColumnsExistence(cs)
+      checkValidJoin(cs)
+    }
+  }
+
+  protected def checkValidJoin(columns: Seq[NamedColumnRef]): Seq[NamedColumnRef] = {
+    val allColumnNames = tableDef.allColumns.map(_.columnName).toSet
+    val regularColumnNames = tableDef.regularColumns.map(_.columnName).toSet
+    val partitionKeyColumnNames = tableDef.partitionKey.map(_.columnName).toSet
+    val joinColumnNames = columns.map(_.columnName).toSet
+
+    def checkSingleColumn(column: NamedColumnRef) = {
+      if (regularColumnNames.contains(column.columnName))
+        throw new IllegalArgumentException(s"Can't pushdown join on column $column because it is not part of the PRIMARY KEY")
+      column
+    }
+
+    //Make sure we have all of the clustering indexes between the 0th position and the max requested in the join
+    val chosenClusteringColumns = tableDef.clusteringColumns
+      .filter(tableCol => columns.exists(joinCol => tableCol.columnName == joinCol.columnName))
+    if (chosenClusteringColumns.size != 0) {
+      val chosenClusteringIndexes = chosenClusteringColumns.flatMap(col => col.componentIndex)
+      val maxIndex = chosenClusteringIndexes.max
+      val requiredIndexes = 0 to maxIndex
+      val missingIndexes = requiredIndexes.toSet -- chosenClusteringIndexes.toSet
+      if (missingIndexes.size != 0) {
+        val maxCol = tableDef.clusteringColumns
+          .filter(cc => cc.componentIndex.get == maxIndex)
+          .map(_.columnName).mkString(", ")
+        val otherCols = tableDef.clusteringColumns
+          .filter(cc => missingIndexes.contains(cc.componentIndex.get))
+          .map(_.columnName).mkString(", ")
+        throw new IllegalArgumentException(s"Can't pushdown join on column ${maxCol} without also specifying [ ${otherCols} ]")
```

Comment:
No need to enclose single variables in string interpolation in {}, $ alone is enough.